libdispatch/dinfermodel.c

```diff
@@ -0,0 +1,1024 @@
+	nclistpush(list,s);
+	if(*endp == '\0') break;
+	p = endp+1;
+    }
+
+done:
+    return check(stat);
+}
+
+/* Given a mode= argument, and the mode flags,
+   infer the iosp part of the model */
+static int
+extractiosp(NClist* modeargs, int cmode, NCmodel* model)
+{
+    int stat = NC_NOERR;
+    struct IOSPS* io = iosps;
+
+    assert(model->iosp == 0);
+    for(;io->tag;io++) {
+	int i;
+	for(i=0;i<nclistlength(modeargs);i++) {
+	    const char* p = nclistget(modeargs,i);
+	    if(strcmp(p,io->tag)==0) {
+		conflictset(MIO,model->iosp,io->iosp);
+		goto done;
+	    }
+	}
+    }
+done:
+    if(model->iosp == 0)
+	model->iosp = (fIsSet(cmode,NC_INMEMORY) ? NC_IOSP_MEMORY:NC_IOSP_FILE);
+    return stat;
+}
+
+/* Given a mode= argument, fill in the matching part of the model; except IOSP */
+static int
+processmodearg(const char* arg, NCmodel* model)
+{
+    int stat = NC_NOERR;
+    struct FORMATMODES* format = formatmodes;
+    for(;format->tag;format++) {
+	if(strcmp(format->tag,arg)==0) {
+	    conflictset(MF,model->format,format->format);
+	    conflictset(MI,model->impl,format->impl);
+	}
+    }
+done:
+    return check(stat);
+}
+
+/* Search singleton list */
+static int
+issingleton(const char* tag)
+{
+    const char** p;
+    for(p=modesingles;*p;p++) {
+	if(strcmp(*p,tag)==0) return 1;
+    }
+    return 0;
+}
+
+/* If we have a url, see if we can determine DAP */
+static int
+NC_dapinfer(NClist* modeargs, NCmodel* model)
+{
+    int stat = NC_NOERR;
+    int i;
+
+    /* 1. search modeargs for indicators */
+    for(i=0;i<nclistlength(modeargs);i++) {
+	const char* arg = nclistget(modeargs,i);
+	if(strcasecmp(arg,"bytes")==0
+	   || strcasecmp(arg,"zarr")==0) {
+	    /* Ok, we know this is not DAP, so give up */
+	    return stat;
+	}
+	if(strcasecmp(arg,"dap2")==0) {
+	    model->format = NC_FORMAT_NC3;
+	    model->iosp = NC_IOSP_DAP2;
+	    model->impl = NC_FORMATX_DAP2;
+	} else if(strcasecmp(arg,"dap4")==0) {
+	    model->format = NC_FORMAT_NETCDF4;
+	    model->iosp = NC_IOSP_DAP4;
+	    model->impl = NC_FORMATX_DAP4;
+	}
+    }
+    /* Ok, we have a URL, but no tags to tell us what it is, so assume DAP2 */
+    if(model->impl == 0) {
+	model->format = NC_FORMAT_NC3;
+	model->iosp = NC_IOSP_DAP2;
+	model->impl = NC_FORMATX_DAP2;
+    }
+    return stat;
+}
+
+/*
+Infer from the mode
+only call if iscreate or file is not easily readable.
+*/
+static int
+NC_omodeinfer(int cmode, NCmodel* model)
+{
+    int stat = NC_NOERR;
+
+    /* If no format flags are set, then use default */
+    if(!fIsSet(cmode,NC_FORMAT_ALL))
+	conflictset(MF,model->format,nc_get_default_format());
+
+    /* Process the cmode; may override some already set flags */
+    if(fIsSet(cmode,NC_64BIT_OFFSET)) {
+	conflictset(MF,model->format,NC_FORMAT_64BIT_OFFSET);
+    }
+    if(fIsSet(cmode,NC_64BIT_DATA)) {
+	conflictset(MF,model->format,NC_FORMAT_64BIT_DATA);
+    }
+    if(fIsSet(cmode,NC_NETCDF4)) {
+	conflictset(MF,model->format,NC_FORMAT_NETCDF4);
```

Comment:
I'm looking at an issue where passing `NC_NOWRITE | NC_64BIT_OFFSET | NC_NETCDF4 | NC_CLASSIC_MODEL` flags worked in 4.6.2, but fails in 4.7.0 in this function. It seems to be detecting that we have the `NC_FORMAT_64BIT_OFFSET` as the format and then fails here because the format has already been set. Is this intended? Should it have worked before?